### PR TITLE
feat(mTLS): adds mTLS support to dataplane api-server

### LINKED ISSUE #50: mTLS for dataplane API
## Problem Statement

We want to have [mutual TLS authentication](https://grpc.io/docs/guides/auth/) for the dataplane GRPC API.

## Acceptance Criteria

-  [ ] mTLS is automatically configured for encrypting _and authenticating_ controlplane to dataplane communications

### PR DESCRIPTION
**Pull request by EandrewJones**
Closes #50.

### Questions
I am not sure how best to add tests for this. Unit testing the config trivially tests the behavior of Opt which we know works. Unit testing `setup_tls` forces me to mock `read_to_string` in the absence of an actual file to read in and all the ways I've read about doing that in rust feel really kludgie / dirty up the code with dependency injections, etc.

1. Is this all we need to add mTLS support for the api-server?
2. Is there some sort of API contract we need to maintain for this such that I can write a conformance or integration test?
3. What documentation would we like to see for this?

### CODE REVIEW

File: dataplane/api-server/src/config.rs
Comment (aryan9600) [inline dataplane/api-server/src/config.rs:15]: i might be lacking the required context but is there a reason this is not `server_certificate_authority_root_path`?

File: dataplane/api-server/src/config.rs
Comment (aryan9600) [inline dataplane/api-server/src/config.rs:25]: is this solely done to avoid a move? asking because i'm not a big fan of inline functions unless they're essential. this can be just its own function that accepts a vector. please correct me if i'm wrong.

File: dataplane/api-server/src/lib.rs
Comment (aryan9600) [inline dataplane/api-server/src/lib.rs:57]: lets bubble this error up instead? `async fn start()` already returns a `Result`, so we can just return it up there.

File: dataplane/api-server/src/config.rs
Comment (aryan9600) [inline dataplane/api-server/src/config.rs:20]: the below two properties don't seem to be actually used anywhere?

File: dataplane/api-server/src/config.rs
Comment (aryan9600) [inline dataplane/api-server/src/config.rs:24]: i couldn't find where this is being called. please ignore if clap calls this automatically for us.

File: dataplane/loader/src/main.rs
Comment (aryan9600) [inline dataplane/loader/src/main.rs:25]: maybe?

```suggestion
    grpc_tls_config: GrpcConfig,
```

File: dataplane/api-server/src/config.rs
Comment (aryan9600) [inline dataplane/api-server/src/config.rs:10]: maybe?

```suggestion
pub struct GrpcTLSConfig {
```

**Comment by aryan9600** [comment]
> Unit testing setup_tls forces me to mock read_to_string in the absence of an actual file to read in

how about generating certs during the test, writing them to temp files and then passing those to `setup_tls`?

> 1. Is this all we need to add mTLS support for the api-server?

this looks good. we can test it using https://github.com/islishude/grpc-mtls-example/blob/main/cmd/client/main.go.

> 2.  Is there some sort of API contract we need to maintain for this such that I can write a conformance or integration test?

we don't have an integration test for the current grpc server. i guess an integration test for the api server would involve spinning up the server and calling each method from an independent client? for testing mTLS specifically, i think we can just get away with only testing one method, since the thing we want to test is actually the TLS connection. something, like the answer to your first question, but coded in a script.

>  3. What documentation would we like to see for this?

we don't really have any docs apart from code comments. i think code comments for now is fine. documentation is an entirely separate conversation that we need to have soon.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:15]: Updating name.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:25]: No moving at play here. I generally agree nested functions suck and avoid them. In this case, I felt it wasn't that big of a deal because it's not intended to be used anywhere outside of the validate method. I tend to feel extremely tight coupling is the only excuse for nested functions.

Either way, I moved it out.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:24]: I swore in my reading, I saw clap would use this method for us. But, alas, I cannot find it in docs. I added an explicit call for this now.

If you know of a better way to get clap to perform validation at parse time, please let me known.

File: dataplane/loader/src/main.rs
Comment (EandrewJones) [reply dataplane/loader/src/main.rs:25]: Good call out.

I have changed the name of the struct to match the function argument: `GrpcConfig` --> `TlsConfig`. There's nothing about the config that makes it unique to Grpc so I don't know why I named it that way originally.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:10]: Changed to TLSConfig

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:20]: correct. These are only relevant client side or if this server also acted as a client w.r.t to a different server.

Removing.

File: dataplane/api-server/src/lib.rs
Comment (EandrewJones) [reply dataplane/api-server/src/lib.rs:57]: K.

**Comment by EandrewJones** [comment]
> how about generating certs during the test, writing them to temp files and then passing those to `setup_tls`?

Added.

> > 1. Is this all we need to add mTLS support for the api-server?
> 
> this looks good. we can test it using https://github.com/islishude/grpc-mtls-example/blob/main/cmd/client/main.go.

If we're not treating this as an integration test, where should we add this so it runs in our pipeline? We need the dataplane server running so we can mimic the client, so it seems like it should go into the integration test pipeline (`tests/integration`). However, the makefile states these tests are deprecated, so should I really be adding new tests to the suite?

> > 3. What documentation would we like to see for this?
> 
> we don't really have any docs apart from code comments. i think code comments for now is fine. documentation is an entirely separate conversation that we need to have soon.

Okay.

File: dataplane/api-server/src/config.rs
Comment (shaneutt) [inline dataplane/api-server/src/config.rs:18]: So in general I've seen a lot of `Option<T>` fields in struct attributes over the years and that's fine in general. What's interesting in this case is that you're not trying to express that these fields are ultimately optional, but instead that they are optional situationally based on the values of other fields. In effect, this means that you have a couple of different _variants_ of a `TLSConfig` that are acceptable.

Given that notion, and the fact that I immediate thought "variants", I think you should consider whether an [Enumeration](https://doc.rust-lang.org/rust-by-example/custom_types/enum.html) might actually be a better fit here, where we would have a `TLSConfig::ServerOnlyTLS` variant and a `TLSConfig::MutualTLS`.

Now... the only problem with this is that I think the standards really don't accept `ServerOnlyTLS` as official nomenclature everywhere they often just say `TLS`, so... that might require some finesse but either way just think this over and let me know what you think.

File: dataplane/api-server/src/config.rs
Comment (shaneutt) [inline dataplane/api-server/src/config.rs:51]: If we go with the `Enum` approach outlined above, then none of this Validation would be needed, but we could create [associated functions](https://doc.rust-lang.org/stable/book/ch05-03-method-syntax.html#associated-functions) to [construct](https://rust-unofficial.github.io/patterns/idioms/ctor.html) these variants.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:18]: Why not just `TLSConfig::TLS` and `TLSConfig::MutualTLS`? Those correspond to the `TLS` and `mTLS` acronyms, respectively, which are standard afaik.

File: dataplane/api-server/src/config.rs
Comment (EandrewJones) [reply dataplane/api-server/src/config.rs:51]: Is this even necessary? Shouldn't clap handle this for us during parsing? The only reason I was doing this was to enforce the conditional dependencies between fields. The enum renders that irrelevant.

Can't we just pass `opt.tls_config` down to the `setup_tls` function and match on the variant?

File: dataplane/api-server/src/config.rs
Comment (shaneutt) [reply dataplane/api-server/src/config.rs]: I'm good with those names :+1:

File: dataplane/api-server/src/config.rs
Comment (shaneutt) [reply dataplane/api-server/src/config.rs]: Not necessary no, just a thought depending on how you'd like this to end up looking. But the spirit behind it was in-line with the second thing you were suggesting: once we use an enum "validation" does effectively just get outmoded by matching on the variants and applying the appropriate functionality thereafter. :+1:

**Comment by EandrewJones** [comment]
DO NOT MERGE. Still needs a test.

**Comment by EandrewJones** [comment]
I pushed up a go test that creates certs, spins up the dataplane, and dials the grpc w/ mTLS from a client. The test passes, but I suspect it's a false positive because the dataplane exits almost immediately on my machine due to an error:

```
Error: map error: failed to create map `AYA_LOGS` with code -1

Caused by:
    0: failed to create map `AYA_LOGS` with code -1
    1: Operation not permitted (os error 1)
Error: map error: failed to create map `GATEWAY_INDEXES` with code -1

Caused by:
    0: failed to create map `GATEWAY_INDEXES` with code -1
    1: Operation not permitted (os error 1)
```
I only care about having a valid test at this point and adding it to our CI.

File: test/integration/dataplane_mtls_test.go
Comment (EandrewJones) [inline test/integration/dataplane_mtls_test.go:162]: Not using KTF to spin up the dataplane. I believe that's a problem.

Comment (shaneutt) [reply test/integration/dataplane_mtls_test.go:162]: Right, we use KTF specifically for MetalLB which we use for IPAM for our (bpf hijacked) services right now. So for a complete testing environment we're still a bit bound to KTF (or you can manually reproduce the same environment with your own metallb configmap).

**Comment by EandrewJones** [comment]
Does the golang controlplane archival have any impact on this? Or do I just
need to stand up KTF before I run these tests.

Best

Evan Jones
Website: www.ea-jones.com

**Comment by shaneutt** [comment]
> Does the golang controlplane archival have any impact on this? Or do I just need to stand up KTF before I run these tests?

The archival shouldn't have an impact: there's a `make build.cluster` target that sends the right flags (and will automatically install `ktf` for you). The Golang integration tests actually were left in `main` despite removing all the other Go code (since the tests in still have value and help us validate our Rust changes until we replace them with Rust tests).

**Comment by EandrewJones** [comment]
@shaneutt @aryan9600 This should be ready for final review. I finally have passing tests. Upon `lgtm`, I'll squash my commits.

**Comment by EandrewJones** [comment]
Rebased.

Best

Evan Jones
Website: www.ea-jones.com

**Comment by EandrewJones** [comment]
Looks like changes to the docker-build on main break this branches docker-build now. Looks like it's due to the branch name breaking docker syntax (can't include '#' is my guess).

I won't have time to touch this this week. :(

**Comment by shaneutt** [comment]
Ok! No worries, no rush.

**Comment by EandrewJones** [comment]
You still can review. I don't recall what I've done since you last
reviewed. Maybe the enum variants and the test?

File: dataplane/api-server/src/lib.rs
Comment (shaneutt) [inline dataplane/api-server/src/lib.rs:41]: If it's OK with you I would like to put a finer point on this comment to make it clear that no dataplane functionality is enabled through this unencrypted endpoint:

```suggestion
    // Public server without TLS (healthchecks ONLY)
```

File: dataplane/api-server/src/lib.rs
Comment (shaneutt) [inline dataplane/api-server/src/lib.rs:71]: I always appreciate when we link to source materials and relevant external documentation, ty :+1:

File: dataplane/api-server/tests/test_setup_tls.rs
Comment (shaneutt) [inline dataplane/api-server/tests/test_setup_tls.rs:87]: For completeness it would be ideal to cover the negative inverse cases of both these positive test cases. That is to say: these tests currently build up to an expected success result and verify it, we should _also_ cover the failure cases and verify that we get the expected error result.

File: dataplane/loader/src/main.rs
Comment (shaneutt) [inline dataplane/loader/src/main.rs:64]: Your taking the time to add documentation here is greatly appreciated :bow: :+1:

File: dataplane/loader/src/main.rs
Comment (shaneutt) [inline dataplane/loader/src/main.rs:105]: :+1:

File: test/integration/dataplane_mtls_test.go
Comment (shaneutt) [inline test/integration/dataplane_mtls_test.go:1]: I'm not going to say no to adding Golang tests here because this PR review has gone on long enough already and I appreciate your efforts and would prefer to get this in and focus on iterating than to try and stop and do a major rework again now. That said, I felt it would be remiss of me not to point out that we'll likely need to rewrite this into Rust tests at some point so just keep that in mind! Feel free to consider this comment resolved at your discretion.

File: test/integration/dataplane_mtls_test.go
Comment (shaneutt) [inline test/integration/dataplane_mtls_test.go:44]: Is it worth pointing out here (for the future code reader) that this test is verifying our mTLS configuration works?

```suggestion
// TestGRPCClientWithMTLS tests the client connectivity to the dataplane's gRPC API server. Mutual TLS (mTLS)
// is required to connect to the dataplane API server, and so as a part of this test we create a client certificate
// using the same CA the server used in order to establish mutual trust.
func TestGRPCClientWithMTLS(t *testing.T) {
```

Maybe a bit too wordy, so feel free to ignore this specific suggestion, but I think a breadcrumb here that let's future readers better understand that mTLS configuration is effectively being tested here would be good.

File: test/integration/dataplane_mtls_test.go
Comment (shaneutt) [inline test/integration/dataplane_mtls_test.go:76]: There's a bunch of optionality present in here, in that it will try and generate certificates if not present, e.t.c.. However (if I'm not misunderstanding) we actually currently expect for all runs of this test that all values are present? If so, and because this is a private function only used here for this purpose, we may actually just want to _fail_ when things are not present instead as this will help us more quickly realize when we've broken our build system. Wdyt?

**Review by shaneutt — APPROVED** [review]
Thank you so much for all your hard work making this happen.

I approve in that I think we're just about ready to go and I don't see any major blocking issues, however I do have some comments I would like to see resolved before we merge please. Also we gotta figure out the CI issue, but in any case, I think we're really close. Thank you!

File: Makefile
Comment (EandrewJones) [reply Makefile:261]: I am happy mpt to remove them since this test will mostly run in ephemeral CI environments. But I think we _should keep_ the targets for creating them. We want to have an automated way to create self-signed certs.

File: dataplane/loader/src/main.rs
Comment (EandrewJones) [reply dataplane/loader/src/main.rs:105]: I don't love the way rust destructures (or doesn't, rather) the config when printing it, but that's a small nit I can live with for now.

File: test/integration/dataplane_mtls_test.go
Comment (EandrewJones) [reply test/integration/dataplane_mtls_test.go:1]: Yeah, I'm aware. I wrote these in go out of laziness; I drew on the precedent of integration tests written in golang.

File: test/integration/dataplane_mtls_test.go
Comment (EandrewJones) [reply test/integration/dataplane_mtls_test.go:78]: This function does not generate certs if they are not absent, it just reads in the certs and converts them into the tls.config struct that grpc expects.

However, it doesn't abort early if paths to CertFile, KeyFile or CAFile are not present. I can add those.

**Comment by EandrewJones** [comment]
@shaneutt I addressed all your comments and fixed the docker-build naming issue by switching the workflow to use the PR number instead of branch name since it's guaranteed to comply with docker's naming rules.

The last thing I need to do is squash. Is there a way for me to squash and re-push without triggering the docker-build action? It's extremely expensive/long-running.

File: Makefile
Comment (shaneutt) [reply Makefile:261]: Right, exactly what I was getting at. Keep the _targets_ but don't do cleanup automatically here. :+1:

**Review by shaneutt — APPROVED** [review]
/approve
/unhold

/cc @aryan9600

**Comment by shaneutt** [comment]
Oh wow yeah 70m build time... holy cow.

**Review by shaneutt — APPROVED** [review]
We could always spend more time here, but given the project is still nascent and we do not have end-users for breakages to impact, let's merge this and start using it!

Thanks for all your hard work @EandrewJones :bow: 

/lgtm

**Pull request merged**
